modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotStorage.scala

```diff
@@ -65,3 +73,68 @@ object GlobalSnapshotStorage {
   case class InvalidGlobalSnapshotChain(expected: GlobalSnapshotLink, actual: GlobalSnapshotLink) extends NoStackTrace
 
 }
+
+object GlobalSnapshotStorageV2 {
+
+  abstract class RingBuffer[F[_], A](capacity: PosInt) {
+    def prepend(a: A): F[Unit]
+    def onEviction(a: A): F[Unit]
+  }
+
+  def make[F[_]: Async: KryoSerializer](
+    genesis: GlobalSnapshot,
+    inMemoryCapacity: PosInt
+  ): F[GlobalSnapshotStorageV2[F]] = {
+    def mkCacheQueue = Queue.unbounded[F, GlobalSnapshot]
+    def mkHeadRef = Ref.of[F, GlobalSnapshot](genesis)
+    def mkCache = MapRef.ofSingleImmutableMap[F, SnapshotOrdinal, GlobalSnapshot](Map.empty)
+    def mkOffloadQueue = Queue.unbounded[F, SnapshotOrdinal]
+
+    (mkCacheQueue, mkHeadRef, mkCache, mkOffloadQueue).mapN { make(_, _, _, _, inMemoryCapacity) }.flatten
+  }
+
+  def make[F[_]: Async: KryoSerializer](
+    cacheQueue: Queue[F, GlobalSnapshot],
+    headRef: Ref[F, GlobalSnapshot],
```

Comment:
No, I've started with SignallingRef but the documentation clearly states that it can skip some values and if all values matters, use Queue or Channel. 

Discrete returns only the last element on SignallingRef, continuous mode returns last element on each Pull (until changed, aka 1, 1, 1, 2, 2, 2, 2, 3, ...)

Reply:
Ok, good to know that!